stdlib/json.gr

```diff
@@ -0,0 +1,2308 @@
+/**
+ * @module Json: JSON (JavaScript Object Notation) parsing and printing.
+ */
+
+import String from "string"
+import Char from "char"
+import List from "list"
+import Option from "option"
+import Number from "number"
+import WasmI32 from "runtime/unsafe/wasmi32"
+import WasmI64 from "runtime/unsafe/wasmi64"
+import WasmF32 from "runtime/unsafe/wasmf32"
+import WasmF64 from "runtime/unsafe/wasmf64"
+import NumberUtils from "runtime/numberUtils"
+import Tags from "runtime/unsafe/tags"
+import Numbers from "runtime/numbers"
+import Memory from "runtime/unsafe/memory"
+import Buffer from "buffer"
+import { toString as runtimeToString } from "runtime/string"
+import { coerceNumberToWasmI32 } from "runtime/numbers"
+import { tagSimpleNumber } from "runtime/dataStructures"
+
+/**
+ * Data structure representing the result of parsing and the input of printing
+ * JSON.
+ * 
+ * For example this object
+ * ```grain
+ * JSONObject([
+ *   ("currency", JSONString("€")),
+ *   ("price", JSONNumber(99.99)),
+ * ])
+ * ```
+ * is equivalent to the following JSON:
+ * ```json
+ * {"currency":"€","price":99.99}
+ * ```
+ * 
+ * This data structure is semantically equivalent to the JSON format allowing
+ * mostly lossless round trips of printing and parsing. Exceptions to this are
+ * white spaces, multiple ways JSON allows escaping characters in strings and
+ * some edge-cases related to Grain's `Number` type.
+ * 
+ * For example parsing the following JSON text will also result in the same
+ * `JSON` object as above.
+ * ```json
+ * {
+ *   "currency": "\u20ac",
+ *   "price": 99.99
+ * }
+ * ```
+ */
+export enum JSON {
+  JSONNull,
+  JSONBoolean(Bool),
+  JSONNumber(Number),
+  JSONString(String),
+  // Here I've chosen to use a List and not an array as a conscious trade of performance for immutability.
+  // I'd love to use a more advanced persistent collection instead, but there currently aren't any in Grain.
```

Comment:
@cician What sort of data-structure do you think would be more appropriate here? This is an important library, so I think it'd be worth road-mapping

Reply:
Maybe it's better if I slim down these two comments and leave just the part about multiple keys. The use of List is just the natural choice in Grain in this case.

It was just me expressing frustration about not having persistent immutable collections, other than the plain old single linked list. What I mean by persistent immutable collections is libraries like https://docs.rs/im/latest/im or https://immutable-js.com. So `JSONArray` could use something like https://docs.rs/im/15.0.0/im/struct.Vector.html.

`JSONObject` is more of a head scratcher because I allow duplicate keys (feel free to enter this rabbit hole with me: https://stackoverflow.com/questions/21832701/does-json-syntax-allow-duplicate-keys-in-an-object). If I did not allow multiple keys, a Map would be more natural, but Gain only has a mutable one.

Since then I decided that it's OK for this API not to present a directly searchable data structure and that this job can be left to a higher level API built on top. No specific plans yet, suggestions welcome.